In [0]:
from google.colab import drive
drive.mount('/content/drive')

Go to this URL in a browser: https://accounts.google.com/o/oauth2/auth?client_id=947318989803-6bn6qk8qdgf4n4g3pfee6491hc0brc4i.apps.googleusercontent.com&redirect_uri=urn%3aietf%3awg%3aoauth%3a2.0%3aoob&response_type=code&scope=email%20https%3a%2f%2fwww.googleapis.com%2fauth%2fdocs.test%20https%3a%2f%2fwww.googleapis.com%2fauth%2fdrive%20https%3a%2f%2fwww.googleapis.com%2fauth%2fdrive.photos.readonly%20https%3a%2f%2fwww.googleapis.com%2fauth%2fpeopleapi.readonly

Enter your authorization code:
··········
Mounted at /content/drive


In [0]:
!cp /content/drive/My\ Drive/Colab\ Notebooks/3-class.csv 3-class.csv

In [0]:
!cp /content/drive/My\ Drive/Colab\ Notebooks/list.txt list.txt

In [0]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt

df = pd.read_csv('3-class.csv')
df.head()

Tweet Polarity
0   ابراهيم_عيسى الوسخ ابن الوسخه كلما حصل حادث ا...      neg
1   اخطر حروب الارض حرب العقيده حسيبك الله ي اول ...      neg
2   اصبحت تقدم برامج عبر الجمعيات الخيريه لايصال ...      neg
3   اعلامنا متمثل في داوودالشريان و روتانا وطقتهم...      neg
4   الاصرار مرتزقه_برنامج_الاصرار بضاعه هالمترديه...      neg

In [0]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 56674 entries, 0 to 56673
Data columns (total 2 columns):
Tweet       56674 non-null object
Polarity    56674 non-null object
dtypes: object(2)
memory usage: 885.7+ KB


In [0]:
df.Polarity.value_counts()

neg     20731
neut    18726
pos     17217
Name: Polarity, dtype: int64

In [0]:
df['Polarity'] = df['Polarity'].map({'neg': 0, 'pos': 1, 'neut': 2})

In [0]:
df[df.Polarity == 0].head(10)

Tweet  Polarity
0   ابراهيم_عيسى الوسخ ابن الوسخه كلما حصل حادث ا...         0
1   اخطر حروب الارض حرب العقيده حسيبك الله ي اول ...         0
2   اصبحت تقدم برامج عبر الجمعيات الخيريه لايصال ...         0
3   اعلامنا متمثل في داوودالشريان و روتانا وطقتهم...         0
4   الاصرار مرتزقه_برنامج_الاصرار بضاعه هالمترديه...         0
5   الاعلام اللبناني يهاجم السعوديه منذ مده بكل ق...         0
6   البرنامج استاجر بعض المشاهير و الهوامير في تو...         0
7   الحمد لله ما احتاج اتعلم من واحد فاشل اخلاقيا...         0
8   الخرج بيض الله وجه محافظ الخرج فهذه القناه تص...         0
9   الرياض السعوديه رسالتي لوزير العمل في حينه عن...         0

In [0]:
df[df.Polarity == 1].head(10)

Tweet  Polarity
39457   ssr true اشكر قناه واشكر الاخوان في منتدى توا...         1
39458              احلام الملكه بانوراما_fm منوره احلام          1
39459          احلى ديو علا الفارس علي الغفيلي في اسبوع          1
39460                           احلى قعده مع احلى رفقات          1
39461                             احلى قعده مع الاصدقاا          1
39462   الحمدلله كانت حفله حلوه وناجحه ومنشكر اهل الم...         1
39463   الكبار برنامج سيساهم في زياده وعي اللاعب السع...         1
39464   امجد يا شاكر يا قرشي نتمنا ليك التوفيق و انشا...         1
39465   انا وحبيبتي الفنانه القديره لطيفه من وسط بيرو...         1
39466   اولا العفو والشكر لكم وجزاكم الله الف خير لكل...         1

In [0]:
df[df.Polarity == 2].head(10)

Tweet  Polarity
20731   FM الاان عن هوايه التطعيس في المنطقه و الاهتم...         2
20732              اقصد برنامج اصرار على قناه تعرض الان          2
20733   الاستاذ محمد الحميدي ضيف في برنامج الثامنه_مع...         2
20734   الان في الطائره واجمل مفاجاه من و لوين بتتوقع...         2
20735   البرنسيسه منال_موسى و سفير الالحان فايز_السعي...         2
20736   الثلاثااالفنانه نوال_الزغبي ضيفه العلاميه وفا...         2
20737   الفيلسوف يوسف الثنيان في برنامج مع الكبار على...         2
20738   اللي له خلق يتابع الساعه 12 30 على mbc برنامج...         2
20739   الليله الاسطوره يوسف الثنيان عبر برنامج الكبا...         2
20740   الليله حلقه نوال_الزغبي في المتاهه على شاشه i...         2

As can be seen from above, the first 20731 entries are of negative class, then the next 18726 tweets have neutral sentiment and after that the final 17217 entries belong to the positive class.

In [0]:
df.Polarity.value_counts()

0    20731
2    18726
1    17217
Name: Polarity, dtype: int64

###Data Preparartion

In [0]:
df['pre_clean_len'] = [len(t) for t in df.Tweet]

In [0]:
from pprint import pprint
data_dict = {
    'polarity': {
        'type': df.Polarity.dtype,
        'description': 'sentiment class - 0: negative, 1: positive, 2: neutral'
    },
    'tweet': {
        'type': df.Tweet.dtype,
        'description': 'tweet text'
    },
    'pre_clean_len': {
        'type': df.pre_clean_len.dtype,
        'description': 'length of the tweet before cleaning'
    },
    'dataset.shape': df.shape
}

pprint(data_dict)

{'dataset.shape': (56674, 3),
 'polarity': {'description': 'sentiment class - 0: negative, 1: positive, 2: '
                             'neutral',
              'type': dtype('int64')},
 'pre_clean_len': {'description': 'length of the tweet before cleaning',
                   'type': dtype('int64')},
 'tweet': {'description': 'tweet text', 'type': dtype('O')}}


* In some tweets the HTML encodings are not converted to 
text, so we'll convert the HTML decoding to general text. We can use beautiful soup for decoding the HTML encodings.

* We also have to remove the '@' charcter mentions as it is not relevant to us.

* We'll also remove the URL links as they contain little significance for our task.

* We have to remove hashtag and numbers from the tweets too.

In [0]:
# Data cleaning function definition

from nltk.tokenize import WordPunctTokenizer
tok = WordPunctTokenizer()

In [0]:
import re
from bs4 import BeautifulSoup

pat1 = r'@[A-Za-z0-9_]+'
pat2 = r'https?://[^ ]+'
combined_pat = r'|'.join((pat1, pat2))
www_pat = r'www.[^ ]+'
arabic_num_pat = '[٠١٢٣٤٥٦٧٨٩]'

def tweet_cleaner(text):
  soup = BeautifulSoup(text, 'lxml')
  souped = soup.get_text()
  try:
        bom_removed = souped.decode("utf-8-sig").replace(u"\ufffd", "?")
  except:
        bom_removed = souped
  stripped = re.sub(combined_pat, '', bom_removed)
  stripped = re.sub(www_pat, '', stripped)
  stripped = re.sub(arabic_num_pat, '', stripped)
  words = [x for x in tok.tokenize(stripped) if len(x) > 1]
  return (" ".join(words)).strip()

In [0]:
clean_tweet_texts = []
for i in range(0, len(df)):
  if( (i+1)%10000 == 0 ):
        print("Tweets %d of %d has been processed"%(i+1,len(df)))                                                                    
  clean_tweet_texts.append(tweet_cleaner(df_copy['Tweet'][i]))

Tweets 10000 of 56674 has been processed
Tweets 20000 of 56674 has been processed
Tweets 30000 of 56674 has been processed
Tweets 40000 of 56674 has been processed
Tweets 50000 of 56674 has been processed


In [0]:
clean_tweet_texts[:10]

['ابراهيم_عيسي الوسخ ابن الوسخه كلما حصل حادث اتهم السعوديه بالارهاب الكلب كان براتب مليون جنيه من سنوي مصر',
 'اخطر حروب الارض حرب العقيده حسيبك الله اول ال راه يم رب عل نا وق ده الع يده وات زا',
 'اصبحت تقدم برامج عبر الجمعيات الخيريه لايصال خبثها اين مسوولي الجمعيه من هذا نطالب_خادم_الحرمين_بايقافmbc',
 'اعلامنا متمثل في داوودالشريان روتانا وطقتهم كيف ترجي من هالاشكال خير همهم الوحيد في الحياه قياده_المراه_للسياره',
 'الاصرار مرتزقه_برنامج_الاصرار بضاعه هالمترديه قناه العهر مزجاه في جميع المجالات',
 'الاعلام اللبناني يهاجم السعوديه منذ مده بكل قبيح ومجموعه تدعم الاعلام في لبنان باقامه برامجها الضخمه فيها',
 'البرنامج استاجر بعض المشاهير الهوامير في تويتر عشان يبررو لهم ويرقعو لقناه برنامج_اصرار',
 'الحمد لله ما احتاج اتعلم من واحد فاشل اخلاقيا همه الشحاذه من علي حساب مواطن_يفضح_برنامج_الاصرار_برعايه_MBC',
 'الخرج بيض الله وجه محافظ الخرج فهذه القناه تصب علي المسلمين سيل من المخالفات الشرعيه والمحرمات تدعو الي الرذيله وتشوه صوره الاسلام',
 'الرياض السعوديه رسالتي لوزير العمل في حينه 

In [0]:
len(clean_tweet_texts)

56674

In [0]:
#normalization
def normalizeArabic(text):
    text = re.sub("[إأٱآا]", "ا", text)
    text = re.sub("ى", "ي", text)
    text = re.sub("ؤ", "و", text)
    text = re.sub("ئ", "ء", text)
    text = re.sub("ة", "ه", text)
    return(text)
for i in range(len(clean_tweet_texts)):
  clean_tweet_texts[i] = normalizeArabic(clean_tweet_texts[i])

In [0]:
clean_df = pd.DataFrame(clean_tweet_texts, columns=['text'])
clean_df['target'] = df.Polarity
clean_df.to_csv('clean_tweet.csv', encoding='utf-8')

In [0]:
!cp /content/drive/My\ Drive/clean_tweet.csv /content/

In [0]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
plt.style.use('fivethirtyeight')

%matplotlib inline
%config InlinBackend.figure_format = 'retina'

In [0]:
csv = 'clean_tweet.csv'
my_df = pd.read_csv(csv, index_col=0)
my_df.head()

text  target
0  ابراهيم_عيسي الوسخ ابن الوسخه كلما حصل حادث ات...       0
1  اخطر حروب الارض حرب العقيده حسيبك الله اول ال ...       0
2  اصبحت تقدم برامج عبر الجمعيات الخيريه لايصال خ...       0
3  اعلامنا متمثل في داوودالشريان روتانا وطقتهم كي...       0
4  الاصرار مرتزقه_برنامج_الاصرار بضاعه هالمترديه ...       0

In [0]:
my_df.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 56674 entries, 0 to 56673
Data columns (total 2 columns):
text      56673 non-null object
target    56674 non-null int64
dtypes: int64(1), object(1)
memory usage: 1.3+ MB


In [0]:
my_df.dropna(inplace=True)
my_df.reset_index(drop=True, inplace=True)
my_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 56673 entries, 0 to 56672
Data columns (total 2 columns):
text      56673 non-null object
target    56673 non-null int64
dtypes: int64(1), object(1)
memory usage: 885.6+ KB


In [0]:
x = my_df.text
y = my_df.target

In [0]:
from sklearn.model_selection import train_test_split
SEED = 666
x_train, x_validation_and_test, y_train, y_validation_and_test = train_test_split(x, y, test_size=.2, random_state=SEED)
x_validation, x_test, y_validation, y_test = train_test_split(x_validation_and_test, y_validation_and_test, test_size=.5, random_state=SEED)

In [0]:
print("Train set has total {0} entries with {1:.2f}% negative, {2:.2f}% positive, {3:.2f}% neutral".format(len(x_train),
                                                                             (len(x_train[y_train == 0]) / (len(x_train)*1.))*100,
                                                                            (len(x_train[y_train == 1]) / (len(x_train)*1.))*100,
                                                                            (len(x_train[y_train == 2]) / (len(x_train)*1.))*100))
print("Validation set has total {0} entries with {1:.2f}% negative, {2:.2f}% positive, {3:.2f}% neutral".format(len(x_validation),
                                                                             (len(x_validation[y_validation == 0]) / (len(x_validation)*1.))*100,
                                                                            (len(x_validation[y_validation == 1]) / (len(x_validation)*1.))*100,
                                                                            (len(x_validation[y_validation == 2]) / (len(x_validation)*1.))*100))
print("Test set has total {0} entries with {1:.2f}% negative, {2:.2f}% positive, {3:.2f}% neutral".format(len(x_test),
                                                                             (len(x_test[y_test == 0]) / (len(x_test)*1.))*100,
                                                                            (len(x_test[y_test == 1]) / (len(x_test)*1.))*100,
                                                                            (len(x_test[y_test == 2]) / (len(x_test)*1.))*100))


Train set has total 45338 entries with 36.59% negative, 30.52% positive, 32.90% neutral
Validation set has total 5667 entries with 36.07% negative, 29.93% positive, 34.00% neutral
Test set has total 5668 entries with 37.01% negative, 29.73% positive, 33.26% neutral


In [0]:
#converting the labels to categorical data

from keras.utils import to_categorical
y_train = to_categorical(y_train)
y_train.shape

(45338, 3)

In [0]:
y_validation = to_categorical(y_validation)

In [0]:
y_validation.shape

(5667, 3)

In [0]:
from tqdm import tqdm
tqdm.pandas(desc="progress-bar")
import gensim
from gensim.models.word2vec import Word2Vec
from gensim.models.doc2vec import TaggedDocument
import multiprocessing
from sklearn import utils

In [0]:
!cp /content/drive/My\ Drive/full_uni_cbow_100_twitter.zip /content/

In [0]:
!cp /content/drive/My\ Drive/full_uni_sg_100_twitter.zip /content/

In [0]:
!unzip full_uni_cbow_100_twitter.zip
!unzip full_uni_sg_100_twitter.zip

Archive:  full_uni_cbow_100_twitter.zip
  inflating: full_uni_cbow_100_twitter.mdl  
  inflating: full_uni_cbow_100_twitter.mdl.trainables.syn1neg.npy  
  inflating: full_uni_cbow_100_twitter.mdl.wv.vectors.npy  
Archive:  full_uni_sg_100_twitter.zip
  inflating: full_uni_sg_100_twitter.mdl  
  inflating: full_uni_sg_100_twitter.mdl.trainables.syn1neg.npy  
  inflating: full_uni_sg_100_twitter.mdl.wv.vectors.npy  


In [0]:
from gensim.models import KeyedVectors
model_ug_cbow = KeyedVectors.load('full_uni_cbow_100_twitter.mdl')
model_ug_sg = KeyedVectors.load('full_uni_cbow_100_twitter.mdl')

/usr/local/lib/python3.6/dist-packages/smart_open/smart_open_lib.py:402: UserWarning: This function is deprecated, use smart_open.open instead. See the migration notes for details: https://github.com/RaRe-Technologies/smart_open/blob/master/README.rst#migrating-to-the-new-open-function
  'See the migration notes for details: %s' % _MIGRATION_NOTES_URL


In [0]:
len(model_ug_cbow.wv.vocab.keys())

1259756

In [0]:
embeddings_index = {}
for w in model_ug_cbow.wv.vocab.keys():
  embeddings_index[w] = np.append(model_ug_cbow.wv[w], 
                                  model_ug_sg.wv[w])
print(f'Found {len(embeddings_index)} word vectors.')

Found 1259756 word vectors.


In [0]:
from keras.preprocessing.text import Tokenizer
from keras.preprocessing.sequence import pad_sequences

tokenizer = Tokenizer(num_words=100000)
tokenizer.fit_on_texts(x_train)
sequences = tokenizer.texts_to_sequences(x_train)

Using TensorFlow backend.


In [0]:
len(tokenizer.word_index)

76792

In [0]:
for x in x_train[:5]:
  print(x)

لماذا_يا_خوله_العنزي انا ماعلي منها هذي حياتها وبلعنتها بس ليش يطلعونها بالتلفزيون انها واجهه مشرفه وكل الشعب يدري انها عار السعوديه
لماذا_يا_خوله_العنزي فالرجاا كل واحد يبلع تبن وينطم والله بيحاسب كل واححد بطلو لقاااااافه
الف مبروك زعماا البطوله 57 للتوضيح بطولات الهلال في سنه واحده بجده والثانيه بلندن والثالثه الرياض مدلع جمهوره بكل مكان الهل
كلمه للتاريخ فارس عوض لماذا الالغاا الهلال_التعاون التعاون_الهلال
تعليق_الدراسه_في_القصيم اتخذ القرار قبل منتصف الليل الامطار تسقط غزاره علي القصيم اي حدث يحدث لا سمح


In [0]:
length = []
for x in x_train:
  length.append(len(x.split()))

In [0]:
max(length)

29

In [0]:
x_train_seq = pad_sequences(sequences, maxlen=35)
print('Shape of data tensor: ', x_train_seq.shape)

Shape of data tensor:  (45338, 35)


In [0]:
x_train_seq[:5]

array([[    0,     0,     0,     0,     0,     0,     0,     0,     0,
            0,     0,     0,    69,    16,    91,   100,    83, 13363,
          374,   212,  7017, 32579,    58,   225, 32580, 16427,   296,
        21592, 32581,   347,   210,  8703,   296,  1868,     6],
       [    0,     0,     0,     0,     0,     0,     0,     0,     0,
            0,     0,     0,     0,     0,     0,     0,     0,     0,
            0,    69,    16,    91,   100, 16428,    22,   161, 32582,
         1820, 21593,    51, 32583,    22, 32584, 11338, 32585],
       [    0,     0,     0,     0,     0,     0,     0,     0,     0,
            0,     0,     0,     0,     0,    29,    30,   577,   243,
          214,  7771,   592,     1,     2,   282,   818,  3549,  9832,
        21594, 16429,    42,  8704,  2105,   297,   496,  1821],
       [    0,     0,     0,     0,     0,     0,     0,     0,     0,
            0,     0,     0,     0,     0,     0,     0,     0,     0,
            0,     0,   

In [0]:
sequences_val = tokenizer.texts_to_sequences(x_validation)
x_val_seq = pad_sequences(sequences_val, maxlen=35)

In [0]:
num_words = 100000
embedding_matrix = np.zeros((num_words, 200))

for word, i in tokenizer.word_index.items():
  if i >= num_words:
    continue
  embedding_vector = embeddings_index.get(word)
  if embedding_vector is not None:
    embedding_matrix[i] = embedding_vector

In [0]:
seed = 3

from keras.models import Sequential
from keras.layers import Dense, Dropout, Flatten
from keras.layers.embeddings import Embedding

In [0]:
from keras.utils import to_categorical
y_train = to_categorical(y_train)
y_train.shape

(45338, 3)

In [0]:
y_validation = to_categorical(y_validation)

In [0]:
!cp /content/full_uni_sg_100_twitter.zip /content/drive/My\ Drive/full_uni_sg_100_twitter.zip

In [0]:
!cp /content/full_uni_cbow_100_twitter.zip /content/drive/My\ Drive/full_uni_cbow_100_twitter.zip

In [0]:
sequences_test = tokenizer.texts_to_sequences(x_test)
x_test_seq = pad_sequences(sequences_test, maxlen=35)
y_test = to_categorical(y_test)

In [0]:
import keras

In [0]:
model_gru = Sequential()
e = Embedding(100000, 200, weights=[embedding_matrix], input_length=35
              , trainable=True)
model_gru.add(e)
# model_gru.add(Flatten())
# model.add(Dense(256, activation='relu'))
model_gru.add(keras.layers.GRU(256, dropout=0.2, recurrent_dropout=0.5))
model_gru.add(Dense(3, activation='softmax'))

model_gru.compile(loss='categorical_crossentropy', optimizer='adam', metrics=['accuracy'])
model_gru.summary()
model_gru.fit(x_train_seq, y_train, 
          validation_data=(x_val_seq, y_validation),
          epochs=5, batch_size=32)


Instructions for updating:
Please use `rate` instead of `keep_prob`. Rate should be set to `rate = 1 - keep_prob`.
Model: "sequential_6"
_________________________________________________________________
Layer (type)                 Output Shape              Param #   
embedding_6 (Embedding)      (None, 35, 200)           20000000  
_________________________________________________________________
gru_2 (GRU)                  (None, 256)               350976    
_________________________________________________________________
dense_9 (Dense)              (None, 3)                 771       
Total params: 20,351,747
Trainable params: 20,351,747
Non-trainable params: 0
_________________________________________________________________
Train on 45338 samples, validate on 5667 samples
Epoch 1/5
45338/45338 [==============================] - 97s 2ms/step - loss: 0.6504 - acc: 0.7265 - val_loss: 0.5429 - val_acc: 0.7780
Epoch 2/5
45338/45338 [==============================] - 95s 2ms/step -

In [0]:
model_lstm = Sequential()
e = Embedding(100000, 200, weights=[embedding_matrix], input_length=35
              , trainable=True)
model_lstm.add(e)
# model_lstm.add(Flatten())
# model.add(Dense(256, activation='relu'))
model_lstm.add(keras.layers.LSTM(256, dropout=0.2, recurrent_dropout=0.5))
model_lstm.add(Dense(3, activation='softmax'))

model_lstm.compile(loss='categorical_crossentropy', optimizer='adam', metrics=['accuracy'])
model_lstm.summary()
model_lstm.fit(x_train_seq, y_train, 
          validation_data=(x_val_seq, y_validation),
          epochs=5, batch_size=32)

Model: "sequential_7"
_________________________________________________________________
Layer (type)                 Output Shape              Param #   
embedding_7 (Embedding)      (None, 35, 200)           20000000  
_________________________________________________________________
lstm_1 (LSTM)                (None, 256)               467968    
_________________________________________________________________
dense_10 (Dense)             (None, 3)                 771       
Total params: 20,468,739
Trainable params: 20,468,739
Non-trainable params: 0
_________________________________________________________________
Train on 45338 samples, validate on 5667 samples
Epoch 1/5
45338/45338 [==============================] - 116s 3ms/step - loss: 0.6200 - acc: 0.7401 - val_loss: 0.5291 - val_acc: 0.7815
Epoch 2/5
45338/45338 [==============================] - 112s 2ms/step - loss: 0.4703 - acc: 0.8102 - val_loss: 0.5236 - val_acc: 0.7964
Epoch 3/5
45338/45338 [==========================

In [0]:
model_bi_gru = Sequential()
e = Embedding(100000, 200, weights=[embedding_matrix], input_length=35
              , trainable=True)
model_bi_gru.add(e)
# model_gru.add(Flatten())
# model.add(Dense(256, activation='relu'))
model_bi_gru.add(keras.layers.Bidirectional(keras.layers.GRU(256, dropout=0.2, recurrent_dropout=0.5)))
model_bi_gru.add(Dense(3, activation='softmax'))

model_bi_gru.compile(loss='categorical_crossentropy', optimizer='adam', metrics=['accuracy'])
model_bi_gru.summary()
model_bi_gru.fit(x_train_seq, y_train, 
          validation_data=(x_val_seq, y_validation),
          epochs=5, batch_size=32)

Model: "sequential_8"
_________________________________________________________________
Layer (type)                 Output Shape              Param #   
embedding_8 (Embedding)      (None, 35, 200)           20000000  
_________________________________________________________________
bidirectional_1 (Bidirection (None, 512)               701952    
_________________________________________________________________
dense_11 (Dense)             (None, 3)                 1539      
Total params: 20,703,491
Trainable params: 20,703,491
Non-trainable params: 0
_________________________________________________________________
Train on 45338 samples, validate on 5667 samples
Epoch 1/5
45338/45338 [==============================] - 198s 4ms/step - loss: 0.6393 - acc: 0.7289 - val_loss: 0.5424 - val_acc: 0.7792
Epoch 2/5
45338/45338 [==============================] - 195s 4ms/step - loss: 0.4921 - acc: 0.7999 - val_loss: 0.5207 - val_acc: 0.7923
Epoch 3/5
45338/45338 [==========================

In [0]:
model_bi_lstm = Sequential()
e = Embedding(100000, 200, weights=[embedding_matrix], input_length=35
              , trainable=True)
model_bi_lstm.add(e)
# model_lstm.add(Flatten())
# model.add(Dense(256, activation='relu'))
model_bi_lstm.add(keras.layers.Bidirectional(keras.layers.LSTM(256, dropout=0.2, recurrent_dropout=0.5)))
model_bi_lstm.add(Dense(3, activation='softmax'))

model_bi_lstm.compile(loss='categorical_crossentropy', optimizer='adam', metrics=['accuracy'])
model_bi_lstm.summary()
model_bi_lstm.fit(x_train_seq, y_train, 
          validation_data=(x_val_seq, y_validation),
          epochs=5, batch_size=32)

Model: "sequential_9"
_________________________________________________________________
Layer (type)                 Output Shape              Param #   
embedding_9 (Embedding)      (None, 35, 200)           20000000  
_________________________________________________________________
bidirectional_2 (Bidirection (None, 512)               935936    
_________________________________________________________________
dense_12 (Dense)             (None, 3)                 1539      
Total params: 20,937,475
Trainable params: 20,937,475
Non-trainable params: 0
_________________________________________________________________
Train on 45338 samples, validate on 5667 samples
Epoch 1/5
45338/45338 [==============================] - 239s 5ms/step - loss: 0.6158 - acc: 0.7437 - val_loss: 0.5405 - val_acc: 0.7810
Epoch 2/5
45338/45338 [==============================] - 231s 5ms/step - loss: 0.4722 - acc: 0.8092 - val_loss: 0.5334 - val_acc: 0.7895
Epoch 3/5
45338/45338 [==========================

In [0]:
model_gru.evaluate(x_test_seq, y_test)

5668/5668 [==============================] - 4s 645us/step


[0.5768967728927763, 0.7985179957657021]

In [0]:
model_lstm.evaluate(x_test_seq, y_test)

5668/5668 [==============================] - 4s 719us/step


[0.5916291314401337, 0.7928722653493295]

In [0]:
model_bi_gru.evaluate(x_test_seq, y_test)

5668/5668 [==============================] - 7s 1ms/step


[0.5900985422736755, 0.7939308398023994]

In [0]:
model_bi_lstm.evaluate(x_test_seq, y_test)

5668/5668 [==============================] - 9s 2ms/step


[0.5953280125273392, 0.7964008468595625]

In [0]:
model_st_gru = Sequential()
e = Embedding(100000, 200, weights=[embedding_matrix], input_length=35
              , trainable=True)
model_st_gru.add(e)
# model_st_gru.add(Flatten())
# model.add(Dense(256, activation='relu'))
model_st_gru.add(keras.layers.GRU(64, dropout=0.2, recurrent_dropout=0.5, return_sequences=True))
model_st_gru.add(keras.layers.GRU(128, dropout=0.2))
model_st_gru.add(Dense(3, activation='softmax'))

model_st_gru.compile(loss='categorical_crossentropy', optimizer='adam', metrics=['accuracy'])
model_st_gru.summary()
model_st_gru.fit(x_train_seq, y_train, 
          validation_data=(x_val_seq, y_validation),
          epochs=5, batch_size=32)

Model: "sequential_10"
_________________________________________________________________
Layer (type)                 Output Shape              Param #   
embedding_10 (Embedding)     (None, 35, 200)           20000000  
_________________________________________________________________
gru_4 (GRU)                  (None, 35, 64)            50880     
_________________________________________________________________
gru_5 (GRU)                  (None, 128)               74112     
_________________________________________________________________
dense_13 (Dense)             (None, 3)                 387       
Total params: 20,125,379
Trainable params: 20,125,379
Non-trainable params: 0
_________________________________________________________________
Train on 45338 samples, validate on 5667 samples
Epoch 1/5
45338/45338 [==============================] - 171s 4ms/step - loss: 0.6542 - acc: 0.7232 - val_loss: 0.5794 - val_acc: 0.7632
Epoch 2/5
45338/45338 [==============================

In [0]:
model_st_gru.evaluate(x_test_seq, y_test)

5668/5668 [==============================] - 7s 1ms/step


[0.5617684845194147, 0.7949894142554693]

In [0]:
model_st_lstm = Sequential()
e = Embedding(100000, 200, weights=[embedding_matrix], input_length=35
              , trainable=True)
model_st_lstm.add(e)
# model_st_gru.add(Flatten())
# model.add(Dense(256, activation='relu'))
model_st_lstm.add(keras.layers.LSTM(64, dropout=0.2, recurrent_dropout=0.5, return_sequences=True))
model_st_lstm.add(keras.layers.LSTM(128, dropout=0.2))
model_st_lstm.add(Dense(3, activation='softmax'))

model_st_lstm.compile(loss='categorical_crossentropy', optimizer='adam', metrics=['accuracy'])
model_st_lstm.summary()
model_st_lstm.fit(x_train_seq, y_train, 
          validation_data=(x_val_seq, y_validation),
          epochs=5, batch_size=32)

Model: "sequential_11"
_________________________________________________________________
Layer (type)                 Output Shape              Param #   
embedding_11 (Embedding)     (None, 35, 200)           20000000  
_________________________________________________________________
lstm_3 (LSTM)                (None, 35, 64)            67840     
_________________________________________________________________
lstm_4 (LSTM)                (None, 128)               98816     
_________________________________________________________________
dense_14 (Dense)             (None, 3)                 387       
Total params: 20,167,043
Trainable params: 20,167,043
Non-trainable params: 0
_________________________________________________________________
Train on 45338 samples, validate on 5667 samples
Epoch 1/5
45338/45338 [==============================] - 201s 4ms/step - loss: 0.6419 - acc: 0.7311 - val_loss: 0.5509 - val_acc: 0.7759
Epoch 2/5
45338/45338 [==============================

In [0]:
model_st_lstm.evaluate(x_test_seq, y_test)

5668/5668 [==============================] - 8s 1ms/step


[0.559693810064017, 0.7944601270289343]

#### Trying out the models without modifying the pre-trained weights of the embedding matrix.

In [0]:
import keras

In [0]:
model_gru = Sequential()
e = Embedding(100000, 200, weights=[embedding_matrix], input_length=35
              , trainable=False)
model_gru.add(e)
# model_gru.add(Flatten())
# model.add(Dense(256, activation='relu'))
model_gru.add(keras.layers.GRU(256, dropout=0.2, recurrent_dropout=0.5))
model_gru.add(Dense(3, activation='softmax'))

model_gru.compile(loss='categorical_crossentropy', optimizer='adam', metrics=['accuracy'])
model_gru.summary()
model_gru.fit(x_train_seq, y_train, 
          validation_data=(x_val_seq, y_validation),
          epochs=5, batch_size=32)


Instructions for updating:
Please use `rate` instead of `keep_prob`. Rate should be set to `rate = 1 - keep_prob`.


Model: "sequential_2"
_________________________________________________________________
Layer (type)                 Output Shape              Param #   
embedding_2 (Embedding)      (None, 35, 200)           20000000  
_________________________________________________________________
gru_1 (GRU)                  (None, 256)               350976    
_________________________________________________________________
dense_1 (Dense)              (None, 3)                 771       
Total params: 20,351,747
Trainable params: 351,747
Non-trainable params: 20,000,000
_________________________________________________________________
Instructions for updating:
Use tf.where in 2.0, which has the same broadcast rule as np.where


Train on 45338 samples, validate on 5667 samples
Epoch 1/5
45338/45338 [==============================] - 82s 2ms/step - loss: 0.6658 - acc: 0.7208 - va

In [0]:
model_lstm = Sequential()
e = Embedding(100000, 200, weights=[embedding_matrix], input_length=35
              , trainable=False)
model_lstm.add(e)
# model_lstm.add(Flatten())
# model.add(Dense(256, activation='relu'))
model_lstm.add(keras.layers.LSTM(256, dropout=0.2, recurrent_dropout=0.5))
model_lstm.add(Dense(3, activation='softmax'))

model_lstm.compile(loss='categorical_crossentropy', optimizer='adam', metrics=['accuracy'])
model_lstm.summary()
model_lstm.fit(x_train_seq, y_train, 
          validation_data=(x_val_seq, y_validation),
          epochs=5, batch_size=32)

Model: "sequential_3"
_________________________________________________________________
Layer (type)                 Output Shape              Param #   
embedding_3 (Embedding)      (None, 35, 200)           20000000  
_________________________________________________________________
lstm_1 (LSTM)                (None, 256)               467968    
_________________________________________________________________
dense_2 (Dense)              (None, 3)                 771       
Total params: 20,468,739
Trainable params: 468,739
Non-trainable params: 20,000,000
_________________________________________________________________
Train on 45338 samples, validate on 5667 samples
Epoch 1/5
45338/45338 [==============================] - 96s 2ms/step - loss: 0.6330 - acc: 0.7345 - val_loss: 0.5550 - val_acc: 0.7759
Epoch 2/5
45338/45338 [==============================] - 94s 2ms/step - loss: 0.5361 - acc: 0.7804 - val_loss: 0.5413 - val_acc: 0.7822
Epoch 3/5
45338/45338 [======================

In [0]:
model_bi_gru = Sequential()
e = Embedding(100000, 200, weights=[embedding_matrix], input_length=35
              , trainable=False)
model_bi_gru.add(e)
# model_gru.add(Flatten())
# model.add(Dense(256, activation='relu'))
model_bi_gru.add(keras.layers.Bidirectional(keras.layers.GRU(256, dropout=0.2, recurrent_dropout=0.5)))
model_bi_gru.add(Dense(3, activation='softmax'))

model_bi_gru.compile(loss='categorical_crossentropy', optimizer='adam', metrics=['accuracy'])
model_bi_gru.summary()
model_bi_gru.fit(x_train_seq, y_train, 
          validation_data=(x_val_seq, y_validation),
          epochs=5, batch_size=32)

Model: "sequential_4"
_________________________________________________________________
Layer (type)                 Output Shape              Param #   
embedding_4 (Embedding)      (None, 35, 200)           20000000  
_________________________________________________________________
bidirectional_1 (Bidirection (None, 512)               701952    
_________________________________________________________________
dense_3 (Dense)              (None, 3)                 1539      
Total params: 20,703,491
Trainable params: 703,491
Non-trainable params: 20,000,000
_________________________________________________________________
Train on 45338 samples, validate on 5667 samples
Epoch 1/5
45338/45338 [==============================] - 166s 4ms/step - loss: 0.6541 - acc: 0.7242 - val_loss: 0.5593 - val_acc: 0.7715
Epoch 2/5
45338/45338 [==============================] - 165s 4ms/step - loss: 0.5511 - acc: 0.7734 - val_loss: 0.5421 - val_acc: 0.7750
Epoch 3/5
45338/45338 [====================

In [0]:
model_bi_lstm = Sequential()
e = Embedding(100000, 200, weights=[embedding_matrix], input_length=35
              , trainable=False)
model_bi_lstm.add(e)
# model_lstm.add(Flatten())
# model.add(Dense(256, activation='relu'))
model_bi_lstm.add(keras.layers.Bidirectional(keras.layers.LSTM(256, dropout=0.2, recurrent_dropout=0.5)))
model_bi_lstm.add(Dense(3, activation='softmax'))

model_bi_lstm.compile(loss='categorical_crossentropy', optimizer='adam', metrics=['accuracy'])
model_bi_lstm.summary()
model_bi_lstm.fit(x_train_seq, y_train, 
          validation_data=(x_val_seq, y_validation),
          epochs=5, batch_size=32)

Model: "sequential_5"
_________________________________________________________________
Layer (type)                 Output Shape              Param #   
embedding_5 (Embedding)      (None, 35, 200)           20000000  
_________________________________________________________________
bidirectional_2 (Bidirection (None, 512)               935936    
_________________________________________________________________
dense_4 (Dense)              (None, 3)                 1539      
Total params: 20,937,475
Trainable params: 937,475
Non-trainable params: 20,000,000
_________________________________________________________________
Train on 45338 samples, validate on 5667 samples
Epoch 1/5
45338/45338 [==============================] - 201s 4ms/step - loss: 0.6300 - acc: 0.7351 - val_loss: 0.5762 - val_acc: 0.7653
Epoch 2/5
45338/45338 [==============================] - 199s 4ms/step - loss: 0.5360 - acc: 0.7790 - val_loss: 0.5428 - val_acc: 0.7777
Epoch 3/5
45338/45338 [====================

In [0]:
model_st_gru = Sequential()
e = Embedding(100000, 200, weights=[embedding_matrix], input_length=35
              , trainable=False)
model_st_gru.add(e)
# model_st_gru.add(Flatten())
# model.add(Dense(256, activation='relu'))
model_st_gru.add(keras.layers.GRU(64, dropout=0.2, recurrent_dropout=0.5, return_sequences=True))
model_st_gru.add(keras.layers.GRU(128, dropout=0.2))
model_st_gru.add(Dense(3, activation='softmax'))

model_st_gru.compile(loss='categorical_crossentropy', optimizer='adam', metrics=['accuracy'])
model_st_gru.summary()
model_st_gru.fit(x_train_seq, y_train, 
          validation_data=(x_val_seq, y_validation),
          epochs=5, batch_size=32)

Model: "sequential_6"
_________________________________________________________________
Layer (type)                 Output Shape              Param #   
embedding_6 (Embedding)      (None, 35, 200)           20000000  
_________________________________________________________________
gru_3 (GRU)                  (None, 35, 64)            50880     
_________________________________________________________________
gru_4 (GRU)                  (None, 128)               74112     
_________________________________________________________________
dense_5 (Dense)              (None, 3)                 387       
Total params: 20,125,379
Trainable params: 125,379
Non-trainable params: 20,000,000
_________________________________________________________________
Train on 45338 samples, validate on 5667 samples
Epoch 1/5
45338/45338 [==============================] - 158s 3ms/step - loss: 0.6688 - acc: 0.7154 - val_loss: 0.5699 - val_acc: 0.7639
Epoch 2/5
45338/45338 [=========================

In [0]:
model_st_lstm = Sequential()
e = Embedding(100000, 200, weights=[embedding_matrix], input_length=35
              , trainable=False)
model_st_lstm.add(e)
# model_st_gru.add(Flatten())
# model.add(Dense(256, activation='relu'))
model_st_lstm.add(keras.layers.LSTM(64, dropout=0.2, recurrent_dropout=0.5, return_sequences=True))
model_st_lstm.add(keras.layers.LSTM(128, dropout=0.2))
model_st_lstm.add(Dense(3, activation='softmax'))

model_st_lstm.compile(loss='categorical_crossentropy', optimizer='adam', metrics=['accuracy'])
model_st_lstm.summary()
model_st_lstm.fit(x_train_seq, y_train, 
          validation_data=(x_val_seq, y_validation),
          epochs=5, batch_size=32)

Model: "sequential_7"
_________________________________________________________________
Layer (type)                 Output Shape              Param #   
embedding_7 (Embedding)      (None, 35, 200)           20000000  
_________________________________________________________________
lstm_3 (LSTM)                (None, 35, 64)            67840     
_________________________________________________________________
lstm_4 (LSTM)                (None, 128)               98816     
_________________________________________________________________
dense_6 (Dense)              (None, 3)                 387       
Total params: 20,167,043
Trainable params: 167,043
Non-trainable params: 20,000,000
_________________________________________________________________
Train on 45338 samples, validate on 5667 samples
Epoch 1/5
45338/45338 [==============================] - 191s 4ms/step - loss: 0.6538 - acc: 0.7252 - val_loss: 0.5708 - val_acc: 0.7681
Epoch 2/5
45338/45338 [=========================

In [0]:
sequences_test = tokenizer.texts_to_sequences(x_test)
x_test_seq = pad_sequences(sequences_test, maxlen=35)
y_test = to_categorical(y_test)

In [0]:
model_gru.evaluate(x_test_seq, y_test)

5668/5668 [==============================] - 4s 733us/step


[0.5331313439297188, 0.7870501058574453]

In [0]:
model_lstm.evaluate(x_test_seq, y_test)

5668/5668 [==============================] - 4s 764us/step


[0.5280382797889669, 0.7858151023288638]

In [0]:
model_bi_gru.evaluate(x_test_seq, y_test)

5668/5668 [==============================] - 8s 1ms/step


[0.5287905923270304, 0.7881086803105152]

In [0]:
model_bi_lstm.evaluate(x_test_seq, y_test)

5668/5668 [==============================] - 9s 2ms/step


[0.5275574734115062, 0.791637261820748]

In [0]:
model_st_gru.evaluate(x_test_seq, y_test)

5668/5668 [==============================] - 8s 1ms/step


[0.5173088391439631, 0.7868736767819337]

In [0]:
model_st_lstm.evaluate(x_test_seq, y_test)

5668/5668 [==============================] - 9s 2ms/step


[0.5148168973882348, 0.7935779816513762]